staging/src/k8s.io/apiserver/pkg/admission/plugin/webhook/generic/webhook.go

```diff
@@ -207,6 +210,25 @@ func (a *Webhook) ShouldCallHook(h webhook.WebhookAccessor, attr admission.Attri
 		return nil, matchObjErr
 	}
 
+	matchConditions := h.GetMatchConditions()
+	expressions := make([]cel.ExpressionAccessor, len(matchConditions))
+	for i, matchCondition := range matchConditions {
+		expressions[i] = &MatchCondition{
+			Name: matchCondition.Name,
+			Expression: matchCondition.Expression,
+		}
+	}
+	matcher := NewMatcher(a.filterCompiler.Compile(expressions, false))
+	versionedAttr, err := admission.NewVersionedAttributes(attr, attr.GetKind(), o)
```

Comment:
We should avoid constructing this twice. It's already being constructed by the dispatcher, but that happens after the call to this method. That might mean that this code needs to be pulled out into a separate method, which is a bit awkward, but I'm not sure I see a better way. WDYT?

Reply:
So we could either construct the versionedAttr sooner and pass it to `ShouldCallHook` however that would have it always construct that versionedAttr in this very high traffic path...

Maybe better if we split up `ShouldCallHook` function into smaller functions like `EvaluateNamespaceSelectors`, `EvaluateObjectSelectors`, `EvaluateRules` which can all be called in the dispatcher before the versionedAttr construction and then finally `EvaluateMatchConditions` which could be called after the versionedAttr construction.

First way would be more simple, second way would be more efficient but would also require slightly more duplicate code in the Dispatch() of VHW and MWH.

Reply:
Or door number 3, part of the return from `ShouldCallHook` returns a versionedAttr